src/ports/postgres/modules/deep_learning/madlib_keras_fit_multiple_model.py_in

```diff
@@ -629,149 +794,187 @@ class FitMultipleModel():
+                        {self.is_final_training_call}::BOOLEAN,
+                        {use_caching}::BOOLEAN,
+                        model_in.{self.object_map_col}::BYTEA
+                    )
+                END::BYTEA AS {self.model_weights_col},
+                model_in.{self.model_arch_col},
+                model_in.{self.compile_params_col},
+                model_in.{self.fit_params_col},
+                model_in.{self.object_map_col},
+                model_in.{self.dist_key_col}
+            FROM {self.model_input_tbl} model_in
+                FULL JOIN {source_table} src
+                USING ({self.dist_key_col}) 
+            DISTRIBUTED BY({self.dist_key_col})
+            """.format(dep_var_col=dep_var,
+                       indep_var_col=indep_var,
+                       dep_shape=dep_shape,
+                       ind_shape=ind_shape,
                        use_caching=self.use_caching,
-                       dist_key_col=dist_key_col,
-                       use_gpus=use_gpus,
                        source_table=source_table,
-                       where_clause=where_clause,
                        self=self
                        )
-        plpy.execute(uda_query)
+        )
+
+        try:
+            plpy.execute(udf_query)
+        except plpy.SPIError as e:
+            msg = e.message
+            if not 'TransAggDetail' in msg:
+                raise e
+            e.message, detail = msg.split('TransAggDetail')
+            # Extract Traceback from segment, add to
+            #  DETAIL of error message on coordinator
+            e.args = (e.message,)
+            spidata = list(e.spidata)
+            spidata[1] = detail
+            e.spidata = tuple(spidata)
+            raise e
+
+        DEBUG.print_timing("udf")
+
+        res = plpy.execute("""
+            SELECT {self.mst_key_col} AS mst_key, {self.model_weights_col} IS NOT NULL AS weights
+                FROM {self.model_output_tbl}
+        """.format(self=self))
+        if res:
+            null_msts = len([None for row in res if row['mst_key'] is None])
+            null_weights = len([None for row in res if row['weights'] is False])
+            DEBUG.plpy.info(
+                "{} rows total ({} mst_key=NULL and {} weights=NULL) in model_output table."\
+                    .format(res.nrows(), null_msts, null_weights))
+        else:
+            plpy.error("No rows in output of UDF!")
 
-        update_query = """
-            UPDATE {self.model_output_table}
-            SET {self.model_weights_col} = {self.weights_to_update_tbl}.{self.model_weights_col}
-            FROM {self.weights_to_update_tbl}
-            WHERE {self.model_output_table}.{self.mst_key_col} = {self.weights_to_update_tbl}.{self.mst_key_col}
-        """.format(self=self)
-        plpy.execute(update_query)
+        plpy.execute("DELETE FROM {self.model_output_tbl} WHERE model_weights IS NULL".format(self=self))
```

Comment:
I was also thinking at first that we could do it that way, using a WHERE clause.  But unfortunately, that can only be used to filter the input rows of a query, not the output rows.  The only way to filter output rows I can think of is to make this query a sub-query of an outer query that does the filtering.

In terms of performance, adding a subquery would be pretty similar to the way it is now, I think, except that it would run the filtering operation as a separate slice... that mostly still has to wait for the first one to complete in order to run.  And I think we want to avoid any extra slices for this one, in case it messes up SD/GD.

I also wonder if we could use a GROUP BY with a HAVING clause.  But that might also add another slice, and/or slightly reduce performance... possibly for the same reason as the UDA ran slightly slower.

The DELETE operation is really fast, since it just scans through the rows and marks some as deleted... the only undesirable effect of this is that it leaves the rows fragmented on disk, but since we're about to TRUNCATE AND DROP it anyway, seems like it shouldn't matter.

We could also change the hop query to filter them out with a WHERE clause when it's copying from `model_output_tbl` to `model_input_tbl`.  But then I guess we'd have to do something similar for eval.  And essentially, this would just be another way of doing the same thing as the DELETE accomplishes, telling them both "ignore these rows".

If you have any other ideas, let me know and I can try it.  (Perhaps using WITH keyword?  My impression has always been that WITH is just another way to introduce a separate sub-query / slice, but not sure.)

Reply:
It would be interesting to figure out how much extra time a subquery would add and if there is a better and efficient way to write this query. But that can be done in a future PR.